### PR TITLE
Bosh Re-Implementation

### PR DESCRIPTION
**Pull request by chris-rock**
- bosh server uses Connection
- improved handling with CORS

### CODE REVIEW

**Comment by sonnyp** [comment]
@chris-rock nice.

If it's a bug in node-xmpp-client shouldn't this be merged anyway? If you're busy I could rebase it.

File: lib/bosh/http.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+ * Sessions are stored *in memory!* You wouldn't want that for very
+ * large setups.
+ */
+function BOSHServer(options) {
+    this.options = options ||  {};
+
+    // set default cors properties
+    if (!this.options.cors) {
+        this.options.cors = {
+            origin : '*'
```
Comment (sonnyp) [inline lib/bosh/http.js]: isn't that dangerous?

File: lib/bosh/server.js [hunk 2]
```diff
@@ -1,117 +1,121 @@
...
-    this.sessions = {}
-    this.server = opts.server
-    this.generateSid = rack(opts.sidBits, opts.sidBitsBase, opts.sidBitsExpandBy)
+function BOSHServer(options) {
+    var self = this;
+    this.options = options ||  {};
+
+    if (!options.autostart) {
+        options.autostart = true;
+    }
```
Comment (sonnyp) [inline lib/bosh/server.js]: so if I set ```options.autostart = false ``` autostart will be enabled? :)

File: lib/bosh/server.js [hunk 3]
```diff
@@ -1,117 +1,121 @@
...
+    this.bosh.on('error', this.emit.bind(this, 'error'))
+    this.bosh.on('end', this.emit.bind(this, 'end'))
+
+    // start bosh server
+    if (this.options.autostart) {
+        debug('start Bosh Server');
+        this.port = this.options.port || this.BOSH_PORT;
+
+        // start bosh http port
+        http.createServer(function (req, res) {
```
Comment (sonnyp) [inline lib/bosh/server.js]: I think it would be great to let user pass its own http/https server as argument

File: lib/bosh/http.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    debug('handle http')
+    this.setCorsHeader(res, this.options.cors );
+    if (req.method === 'POST') {
+        this._handlePostRequest(req, res)
+    } else {
+        debug('http method is not POST')
+        res.writeHead(400)
+        res.end()
+    }
+}
```
Comment (sonnyp) [inline lib/bosh/http.js]: OPTIONS needs to be handled as well

File: lib/bosh/http.js (hunk 1 above)
Comment (sonnyp) [inline lib/bosh/http.js]: Also wildcard isn't allowed with ```Access-Control-Allow-Credentials```

File: lib/bosh/session.js [hunk 5]
```diff
@@ -43,18 +53,71 @@ BOSHServerSession.prototype.setOptions = function(opts) {
...
     this.sid = opts.sid || hat()
-    this.nextRid = parseInt(opts.bodyEl.attrs.rid, 10) + 1
+    // add sid to properties
+    this.xmlnsAttrs.sid = this.sid;
+
+
+    this.nextRid = parseInt(opts.bodyEl.attrs.rid, 10)
     this.wait = parseInt(opts.bodyEl.attrs.wait || '30', 10)
     this.hold = parseInt(opts.bodyEl.attrs.hold || '1', 10)
+    this.inQueue = {}
```
Comment (sonnyp) [inline lib/bosh/session.js]: Object.create(null) is faster and you can avoid using hasOwnProperty which is even faster.

An ES6 Map is also a good candidate if available.

File: lib/bosh/session.js [hunk 6]
```diff
@@ -43,18 +53,71 @@ BOSHServerSession.prototype.setOptions = function(opts) {
...
+BOSHServerSession.prototype.resume = function () {};
+BOSHServerSession.prototype.end = function () {};
+
+/**
+ * internal method to emit data to "Connection"
+ */
+BOSHServerSession.prototype.sendData = function (data) {
+    // emit this data to connection
+    debug('emit data: ' + data.toString())
+    this.emit('data', data.toString())
```
Comment (sonnyp) [inline lib/bosh/session.js]: I'm not sure to understand when the data is actually sent to the socket.

File: lib/bosh/session.js [hunk 7]
```diff
@@ -43,18 +53,71 @@ BOSHServerSession.prototype.setOptions = function(opts) {
...
+
+BOSHServerSession.prototype.pause = function () {};
+BOSHServerSession.prototype.resume = function () {};
+BOSHServerSession.prototype.end = function () {};
+
+/**
+ * internal method to emit data to "Connection"
+ */
+BOSHServerSession.prototype.sendData = function (data) {
+    // emit this data to connection
```
Comment (sonnyp) [inline lib/bosh/session.js]: The data is not actually sent here so I think this comment is misleading.

File: lib/bosh/http.js (hunk 1 above)
Comment (chris-rock) [inline lib/bosh/http.js]: I fully agree, this was migrated from the old code.

File: lib/bosh/server.js (hunk 3 above)
Comment (chris-rock) [inline lib/bosh/server.js]: would be a good point

File: lib/bosh/http.js (hunk 1 above)
Comment (sonnyp) [inline lib/bosh/http.js]: Sorry I'll probably have more comments on old migrated code, feel free to ignore them.

File: lib/bosh/session.js [hunk 8]
```diff
@@ -86,36 +150,71 @@ BOSHServerSession.prototype.handleHTTP = function(opts) {
...
+    // handle stream start
+    if (!restart && rid && !sid) {
+        debug('handle stream start')
+        // emulate stream creation for connection
+        this.sendData("<?xml version='1.0' ?>")
+        this.sendData("<stream:stream xmlns='jabber:client' xmlns:stream='http://etherx.jabber.org/streams' to='" + to + "' version='" + xmppv + "'>")
+    }
+
+    // handle stream reset
+    if (opts.bodyEl.attrs['xmpp:restart'] == "true") {
```
Comment (sonnyp) [inline lib/bosh/session.js]: should this be 
```else if (restart) { ... }``` ?

File: lib/bosh/session.js [hunk 9]
```diff
@@ -86,36 +150,71 @@ BOSHServerSession.prototype.handleHTTP = function(opts) {
...
+    var to = opts.bodyEl.attrs.to
+    var restart = opts.bodyEl.attrs['xmpp:restart'];
+    var xmppv = opts.bodyEl.attrs['xmpp:version'];
+
+    // handle stream start
+    if (!restart && rid && !sid) {
+        debug('handle stream start')
+        // emulate stream creation for connection
+        this.sendData("<?xml version='1.0' ?>")
+        this.sendData("<stream:stream xmlns='jabber:client' xmlns:stream='http://etherx.jabber.org/streams' to='" + to + "' version='" + xmppv + "'>")
```
Comment (sonnyp) [inline lib/bosh/session.js]: maybe we should have some function builders for those

File: lib/bosh/http.js (hunk 1 above)
Comment (chris-rock) [inline lib/bosh/http.js]: but its good to track them

File: lib/bosh/server.js (hunk 3 above)
Comment (chris-rock) [inline lib/bosh/server.js]: thats what happens if autostart is false and you would open a https server by yourself and calls bosh.handleHTTP(req,res). Do you have any ideas for improvement?

File: lib/bosh/session.js (hunk 8 above)
Comment (chris-rock) [inline lib/bosh/session.js]: Sorry I missed that. You're right

File: lib/bosh/session.js (hunk 9 above)
Comment (sonnyp) [inline lib/bosh/session.js]: Also, it doesn't work for me, not sure why, investigating. (Using Strophe.js but I don't think it's related)

The client never receives the data.

```
  xmpp:bosh:server handle bosh request +6s
  xmpp:bosh:http handle http +6s
  xmpp:bosh:server handle bosh request +30ms
  xmpp:bosh:http handle http +30ms
  xmpp:bosh:http handle POST request  +1ms
  xmpp:bosh:http got: <body rid="1761048664" xmlns="http://jabber.org/protocol/httpbind" sid="5047debf13f316794ee82efaefc427d9"/> +4ms
  xmpp:bosh:http sid found, reuse existing session +0ms
  xmpp:bosh:server handle bosh request +71ms
  xmpp:bosh:http handle http +66ms
  xmpp:bosh:http handle POST request  +0ms
  xmpp:bosh:http got: <body rid="2399038169" xmlns="http://jabber.org/protocol/httpbind" to="localhost" xml:lang="en" wait="60" hold="1" content="text/xml; charset=utf-8" ver="1.6" xmpp:version="1.0" xmlns:xmpp="urn:xmpp:xbosh"/> +1ms
  xmpp:bosh:http no sid found, create a new session +0ms
  xmpp:bosh:http create a new session +0ms
  xmpp:connection setup socket +0ms
  xmpp:bosh:server bosh connect +4ms
  xmpp:connection use standard socket +3ms
  xmpp:connection setup stream +1ms
  xmpp:bosh:session run workInQueue +0ms
  xmpp:bosh:session handle stream start +0ms
  xmpp:bosh:session emit data: <?xml version='1.0' ?> +1ms
  xmpp:bosh:session emit data: <stream:stream xmlns='jabber:client' xmlns:stream='http://etherx.jabber.org/streams' to='localhost' version='1.0'> +0ms
  xmpp:bosh:session schedule response +0ms
  xmpp:bosh:session run workOutQueue +0ms
  xmpp:bosh:session run workInQueue +0ms
```

File: lib/bosh/http.js (hunk 4 above)
Comment (chris-rock) [inline lib/bosh/http.js]: I agree, the whole CORS handling is not production ready

File: lib/bosh/session.js (hunk 9 above)
Comment (chris-rock) [inline lib/bosh/session.js]: is the test case working on your machine?

File: lib/bosh/session.js (hunk 8 above)
Comment (sonnyp) [inline lib/bosh/session.js]: ```npm test```

Everything but
```
  1) BOSH client/server events should be in the right order for connecting:
     Error: timeout of 2000ms exceeded
      at null.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/grunt-mocha-cli/node_modules/mocha/lib/runnable.js:139:19)
      at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)

  2) BOSH client/server events should ping pong stanza:
     Error: timeout of 2000ms exceeded
      at null.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/grunt-mocha-cli/node_modules/mocha/lib/runnable.js:139:19)
      at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)

  3) BOSH client/server events should close the connection:
     Uncaught AssertionError: ["clientend","clientclose"] deepEqual ["end","disconnect","close","clientend","clientclose"]
      at Client.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/test/bosh.js:147:24)
      at Client.emit (events.js:92:17)
      at BOSHConnection.emit (events.js:92:17)
      at BOSHConnection.processResponse (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/lib/bosh.js:83:14)
      at BOSHConnection.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/lib/bosh.js:126:26)
      at Request._callback (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/lib/bosh.js:184:17)
      at Request.self.callback (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/node_modules/request/request.js:372:22)
      at Request.emit (events.js:98:17)
      at Request.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/node_modules/request/request.js:1317:14)
      at Request.emit (events.js:117:20)
      at IncomingMessage.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/node_modules/request/request.js:1265:12)
      at IncomingMessage.emit (events.js:117:20)
      at _stream_readable.js:944:16
      at process._tickCallback (node.js:442:13)
```

File: lib/bosh/session.js (hunk 9 above)
Comment (sonnyp) [inline lib/bosh/session.js]: ```npm test```

Everything but
```
  1) BOSH client/server events should be in the right order for connecting:
     Error: timeout of 2000ms exceeded
      at null.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/grunt-mocha-cli/node_modules/mocha/lib/runnable.js:139:19)
      at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)

  2) BOSH client/server events should ping pong stanza:
     Error: timeout of 2000ms exceeded
      at null.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/grunt-mocha-cli/node_modules/mocha/lib/runnable.js:139:19)
      at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)

  3) BOSH client/server events should close the connection:
     Uncaught AssertionError: ["clientend","clientclose"] deepEqual ["end","disconnect","close","clientend","clientclose"]
      at Client.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/test/bosh.js:147:24)
      at Client.emit (events.js:92:17)
      at BOSHConnection.emit (events.js:92:17)
      at BOSHConnection.processResponse (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/lib/bosh.js:83:14)
      at BOSHConnection.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/lib/bosh.js:126:26)
      at Request._callback (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/lib/bosh.js:184:17)
      at Request.self.callback (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/node_modules/request/request.js:372:22)
      at Request.emit (events.js:98:17)
      at Request.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/node_modules/request/request.js:1317:14)
      at Request.emit (events.js:117:20)
      at IncomingMessage.<anonymous> (/Users/sonny/Projects/Redbooth/happy/repo/lib/XMPP/node-xmpp-server/node_modules/node-xmpp-client/node_modules/request/request.js:1265:12)
      at IncomingMessage.emit (events.js:117:20)
      at _stream_readable.js:944:16
      at process._tickCallback (node.js:442:13)
```

File: lib/bosh/server.js [hunk 10]
```diff
@@ -1,117 +1,121 @@
...
+    if (!options.autostart) {
+        options.autostart = true;
+    }
+
+    EventEmitter.call(this);
+
+    this.c2s = null;
+    this.availableSaslMechanisms = [Plain];
+
+    this.bosh = new BOSHHttp();
```
Comment (sonnyp) [inline lib/bosh/server.js]: We should pass this.options here.

File: lib/bosh/server.js (hunk 3 above)
Comment (sonnyp) [inline lib/bosh/server.js]: Actually no, except maybe some way to return a middleware so we can support frameworks like express.

example
```app.use('/http-bind/', BOSHServer.middleware()); ```

**Comment by sonnyp** [comment]
I started addressing some comments here
https://github.com/sonnyp/node-xmpp-server/tree/bosh

**Comment by chris-rock** [comment]
@sonnyp very cool. let me know if you need help

**Comment by sonnyp** [comment]
@chris-rock I think there should be an additional level of abstraction for C2SServer and BOSHServer and eventually WebSocketServer (which I'll start soon) so that you don't have to register event handlers on the three of them (if you need the 3) what do you think?

**Comment by chris-rock** [comment]
@sonnyp lets talk first, i wrote a web socket version already. I post them the next days

**Comment by sonnyp** [comment]
@chris-rock awesome

File: lib/bosh/session.js [hunk 11]
```diff
@@ -86,36 +150,71 @@ BOSHServerSession.prototype.handleHTTP = function(opts) {
...
 
-    opts.bodyEl.children.forEach(this.emit.bind(this, 'stanza'))
+    // handle message
+
+    // extract values
+    var rid = opts.bodyEl.attrs.rid
+    var sid = opts.bodyEl.attrs.sid
+    var to = opts.bodyEl.attrs.to
+    var restart = opts.bodyEl.attrs['xmpp:restart'];
+    var xmppv = opts.bodyEl.attrs['xmpp:version'];
```
Comment (sonnyp) [inline lib/bosh/session.js]: ```xmpp:version``` isn't mandatory.

http://xmpp.org/extensions/xep-0206.html#initiate
"The client SHOULD include a 'version' attribute"

If the client doesn't send it, the response includes ```xmpp:version="undefined"```

Comment (sonnyp) [inline lib/bosh/session.js]: also, ```xmpp:version``` not ```version``` (see line https://github.com/node-xmpp/node-xmpp-server/pull/57/files#diff-e951619f73d6b71822ea35ecfb2a1a1eR178)

File: lib/bosh/http.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    })
+    parser.on('error', function(e) {
+        cb(e)
+    })
+}
+
+/**
+ * Sessions are stored *in memory!* You wouldn't want that for very
+ * large setups.
+ */
```
Comment (sonnyp) [inline lib/bosh/http.js]: Well it's not like there is an alternative.

**Comment by chris-rock** [comment]
@sonnyp how is your implementation going forward

**Comment by sonnyp** [comment]
@chris-rock good, we're investigating what we believe is our last issue and I'll send PR.

File: lib/bosh/session.js [hunk 13]
```diff
@@ -86,36 +150,71 @@ BOSHServerSession.prototype.handleHTTP = function(opts) {
...
+        debug('reset stream')
+        // emulate stream restart for connection
+        var stream = "<stream:stream xmlns='jabber:client' xmlns:stream='http://etherx.jabber.org/streams'  to='" + to + "' version='" + xmppv + "'>"
+        this.sendData(stream);
+    }
+
+    opts.bodyEl.children.forEach(function(stanza) {
+        debug('send data: ' + stanza)
+        // extract content
+        self.sendData(stanza.root().children.toString());
```
Comment (sonnyp) [inline lib/bosh/session.js]: Hey @chris-rock I noticed 2 data events were triggered on the socket. After investigating it looks like this is the root cause. Replaced this line with ```self.sendData(stanza.toString());``` and :+1: 
Does it makes sense to you?

File: lib/bosh/session.js [hunk 14]
```diff
@@ -139,34 +243,14 @@ BOSHServerSession.prototype.workOutQueue = function() {
...
     if (this.outQueue.length < 1) {
+        var self = this
         this.connectionTimeout = setTimeout(function() {
-            this.emit('error', new Error('Session timeout'))
-            this.emit('close')
-        }.bind(this), 60 * 1000)
+            self.emit('error', new Error('Session timeout'))
+            self.emit('close')
+        }, 60 * 1000)
     }
```
Comment (sonnyp) [inline lib/bosh/session.js]: @chris-rock the purpose of this is to define an XMPP timeout right? If no XMPP stanza is exchanged between the server and the client within 60s then the connection is closed.

To me it looks like it should be moved out of BOSH related code. What do you think?

Comment (chris-rock) [inline lib/bosh/session.js]: yeah, let's try this. What would be your approach?

File: lib/bosh/session.js (hunk 13 above)
Comment (chris-rock) [inline lib/bosh/session.js]: I assume this was a workaround for the old node-xmpp core. Lets move forward with your version.

File: lib/bosh/session.js (hunk 14 above)
Comment (sonnyp) [inline lib/bosh/session.js]: No idea, I'm gonna remove it for now let's tackle this in a separate issue.

Comment (chris-rock) [inline lib/bosh/session.js]: Alright. Sounds reasonable.

**Comment by sonnyp** [comment]
@chris-rock https://github.com/node-xmpp/node-xmpp-server/commit/2fab75a709241bbcafcccb3c8882f464f81c5495 I pushed my changes to bosh branch we still need to fix the CORS handling

**Comment by chris-rock** [comment]
@sonnyp I have good experience with https://github.com/troygoode/node-cors Would be great to add this. We need this for websockets, too

**Comment by lloydwatkin** [comment]
I thought websockets didn't suffer CORS issues?

**Comment by chris-rock** [comment]
need coffee ;-) this was socket.io ;-)

**Comment by lloydwatkin** [comment]
We've all been there today @chris-rock :)

**Comment by sonnyp** [comment]
@chris-rock is node-cors working with vanilla node.js http server as well?

**Comment by sonnyp** [comment]
"could not parse<stream:stream xmlns:stream="http://etherx.jabber.org/streams" xmlns="jabber:client" version="1.0" id="9a69231adcf634240a1e213f471a0586" from="example.com">"

```javascript
        // TODO, we need to filter the stream element
        children.forEach(function (element) {
            try {
                bodyEl.cnode(ltx.parse(element))
            } catch (err) {
                console.error('could not parse' + element)
            }
        })
```

Not sure what to do about this "TODO, we need to filter the stream element" any idea?

**Comment by chris-rock** [comment]
@sonnyp that cors module is optimized for express. so that kills the idea. Thanks for the feedback

**Comment by chris-rock** [comment]
Problem is that the stream start with <stream> but no </stream> will follow. ltx cannot handle incomplete tags. Should work for now, but I would like to get rid of the error handling there

**Comment by sonnyp** [comment]
@chris-rock yes, my question was how to properly get rid of the console.error?

**Comment by chris-rock** [comment]
we could just call debug()

**Comment by sonnyp** [comment]
I believe this is ready to be merge.
Tests are green using bosh-restart branch of node-xmpp-client (maybe we should merge it first?) the PR is here https://github.com/node-xmpp/node-xmpp-client/pull/91

**Comment by chris-rock** [comment]
rebased onto master

**Comment by chris-rock** [comment]
@sonnyp could you have a look, why bosh fails on node 0.10?

**Comment by chris-rock** [comment]
Once this is going through, I would love to merge it.

**Comment by sonnyp** [comment]
@chris-rock the BOSH restart PR https://github.com/node-xmpp/node-xmpp-client/pull/91 fixes the test (it has been merged)

**Comment by sonnyp** [comment]
@chris-rock BOSH is an acronym so BOSHServer is correct (and still camelcase)

**Comment by sonnyp** [comment]
travis node 0.10 :+1:  merging this

**Pull request merged**